Update Schema Object

webron opened the pull request:
(This replaced PR #741)

Covers:
- Clarifications about JSON Schema property adjustments to the OAS
- Support for `nullable`
- Support for `writeOnly`, removed `required` restriction
- `discriminator` clarifications
- examples cleanup

DavidBiesack commented on versions/3.0.md:
```diff
@@ -2287,25 +2283,32 @@ The following properties are taken directly from the JSON Schema definition and
+
+The following properties are taken from the JSON Schema definition but their definitions were adjusted to the OpenAPI Specification. 
+- type - Value MUST be a string. Multiple types via an array are not supported.
+- allOf - Inline or referenced schema MUST be of a [Schema Object](#schemaObject) and not a standard JSON Schema.
```
Should this be:
"MUST be of a [Schema Object](#schemaObject) or [Schema Reference](#reference-object)" ? (Ditto on the other items in the list.

DavidBiesack replied:
Please elaborate on what "not a standard JSON Schema." means. However, if it means "not a full JSON Schema but the OpenAPI Specification schema", I think we should just drop this since you already refer explicitly to `[Schema Object](#schemaObject)`

webron replied:
It should be the Schema Object, not Schema Reference (which I assume you meant JSON Reference). The Schema Object contains the JSON Reference in it.

The 'standard JSON Schema' is in reference to the initial point that the Schema Object is an extended subset of JSON Schema. I'd rather keep this explicit here due to questions raised in the past.

whitlockjc replied:
I'm with @DavidBiesack on this one, I would expect `allOf` to take either an inline _Schema Object_ or a _JSON Reference_ to a _Schema Object_ defined elsewhere.

whitlockjc replied:
@DavidBiesack I almost responded with a _"+1"_ to your comment but I think the _"Inline or referenced schema"_ as to being either a _Schema Object_ or a _JSON Reference_ to a _Schema Object_ defined elsewhere.  @webron's working seem fine to me.

darrelmiller replied:
I think it is useful to remind people that even when embedding schema objects in `allOf` what is being embedded is not a regular JSON Schema.  I suspect many people don't realize that the schemas inside OpenAPI are not regular JSON Schemas.

whitlockjc replied:
Doesn't linking to a _Schema Object_ do this?

DavidBiesack commented on versions/3.0.md:
```diff
@@ -2287,25 +2283,32 @@ The following properties are taken directly from the JSON Schema definition and
+- oneOf - Inline or referenced schema MUST be of a [Schema Object](#schemaObject) and not a standard JSON Schema.
+- anyOf - Inline or referenced schema MUST be of a [Schema Object](#schemaObject) and not a standard JSON Schema.
+- not - Inline or referenced schema MUST be of a [Schema Object](#schemaObject) and not a standard JSON Schema.
+- items - Value MUST be an object and not an array. Inline or referenced schema MUST be of a [Schema Object](#schemaObject) and not a standard JSON Schema. `items` MUST be present of the `type` is `array`.
```
typo: "*of* the type" s/b "*if* the type"

DavidBiesack commented on versions/3.0.md:
```diff
@@ -2287,25 +2283,32 @@ The following properties are taken directly from the JSON Schema definition and
+- items - Value MUST be an object and not an array. Inline or referenced schema MUST be of a [Schema Object](#schemaObject) and not a standard JSON Schema. `items` MUST be present of the `type` is `array`.
+- properties - Property definitions MUST be a [Schema Object](#schemaObject) and not a standard JSON Schema (inline or referenced).
+- additionalProperties - Value can be boolean or object. If object, definition MUST be a [Schema Object](#schemaObject) and not a standard JSON Schema (inline or referenced).
+- $ref - As a [Reference Object](#referenceObject).
```
suggest change "As a Reference Object" to "A Reference Object"

DavidBiesack commented on versions/3.0.md:
```diff
@@ -2260,18 +2260,14 @@ $ref: 'definitions.yaml#/Pet'
-- format (See [Data Type Formats](#dataTypeFormat) for further details)
+
 - title
-- description ([CommonMark syntax](http://spec.commonmark.org/) can be used for rich text representation)
```
remove parens here (not used elsewhere in OAS when describing CommonMark)

DavidBiesack commented on versions/3.0.md:
```diff
@@ -2287,25 +2283,32 @@ The following properties are taken directly from the JSON Schema definition and
+- properties - Property definitions MUST be a [Schema Object](#schemaObject) and not a standard JSON Schema (inline or referenced).
+- additionalProperties - Value can be boolean or object. If object, definition MUST be a [Schema Object](#schemaObject) and not a standard JSON Schema (inline or referenced).
+- $ref - As a [Reference Object](#referenceObject).
+- description ([CommonMark syntax](http://spec.commonmark.org/) can be used for rich text representation).
```
remove parens here (not used elsewhere in OAS when describing CommonMark)

DavidBiesack commented on versions/3.0.md:
```diff
@@ -2599,7 +2602,7 @@ definitions:
     },
     "Cat": {
-      "description": "A representation of a cat",
+      "description": "A representation of a cat. Note that `Cat` will be used as the discriminator value.",
```
Please add an example:

Below are examples showing an array of Pet objects using `petType` as a discriminator:
```yaml
- name: "Fluffy"
  petType: "Cat"
  huntingSkill: "lazy"
- name: "Rover"
  petType: "Dog"
  packSize: 6
```

```json
[
  {
    "name" : "Fluffy",
    "petType" : "Cat",
    "huntingSkill" : "lazy"
  },
  {
    "name" : "Rover",
    "petType" : "Dog",
    "packSize" : 6
  }
]

```

webron commented:
Updates:
- Updated JSON Schema support from Draft 04 to Draft 05 (or wright-00).
- Following that, updated support for External references instead of Canonical dereferencing (same thing, different name).
- Updated references to JSON Pointer.
- Updated fallback of tools to deal with unknown formats.
- Some wording updates.

Not included:
- Updated example with `discriminator` value overriding. Pending review of the `discriminator` as a whole.

Let me know if I missed anything we discussed on the call today.

whitlockjc commented on versions/3.0.md:
```diff
@@ -144,14 +144,15 @@ By convention, the OpenAPI Specification (OAS) file is named `openapi.json` or `
-Primitive data types in the OAS are based on the types supported by the [JSON-Schema Draft 4](http://json-schema.org/latest/json-schema-core.html#anchor8).
-Models are described using the [Schema Object](#schemaObject) which is a subset of JSON Schema Draft 4.
+Primitive data types in the OAS are based on the types supported by the [JSON-Schema Draft Wright 00](https://tools.ietf.org/html/draft-wright-json-schema-00#section-4.2) (also known as, JSON Schema Draft 05). Note that `integer` as a type is also supported and is defined as a JSON number without a fraction or exponent part. `null` is not supported as a value.
+Models are described using the [Schema Object](#schemaObject) which is an extended subset of JSON Schema Draft 05.
```
Does it read better using _"JSON-Schema Draft Wright 00"_ here since that is what you mentioned prior?  Also, was hyphenating _JSON Schema_ to _JSON-Schema_ intentional in the link text?  If so, what's the rules behind when we hyphenate and when we do not?

webron replied:
The hyphen is a mistake. Will fix. 

I can change it to JSON Schema Draft Wright 00 everywhere, but it's not the exact name. The exact name is draft-wright-json-schema-00 which is awkward to write across the doc.

whitlockjc replied:
Makes sense.  I trust your judgement.

webron replied:
Fixed.

whitlockjc commented on versions/3.0.md:
```diff
@@ -2214,8 +2215,8 @@ definitions:
 
-The Reference Object is a [JSON Reference](http://tools.ietf.org/html/draft-pbryan-zyp-json-ref-02) that uses a [JSON Pointer](http://tools.ietf.org/html/rfc6901) as its value.
-For this specification, only [canonical dereferencing](http://json-schema.org/latest/json-schema-core.html#anchor27) is supported.
+The Reference Object is a [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03) that uses a [JSON Pointer](http://tools.ietf.org/html/rfc6901) as its value.
```
I'm not sure I understand what's being said here.  Is the _Reference Object_ **only** for local references, as in references within the same document?  If that's the case, the value is itself a _JSON Pointer_ as a _JSON Reference_ allows both local and remote references.

darrelmiller replied:
Reference Object is for all references whether local or remote.

whitlockjc replied:
That's what I was hoping for.  So in that case, the _"that uses a JSON Pointer as its value"_ is not accurate is it?  The _JSON Pointer_ version of a `$ref` value is for _local only_ and since a _Reference Object_ is for local and remote references, isn't the _Reference Object_'s `$ref` just a URI per the [JSON References spec](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03#section-3).

webron replied:
You're right, @whitlockjc. The indication of JSON Pointer here is not needed. It was added in 2.0, back in the days where things were still unclear to us all. We now know better - A JSON References, is a JSON Reference, and in *some* cases the value is a JSON Pointer. The actual explanation here is wrong and unneeded. Will remove.

whitlockjc replied:
👍

webron replied:
Completely reworked JSON Reference, see main comment.

darrelmiller commented:
:shipit:

webron commented:
A few last minute updates to this PR:
- Updated the naming of JSON Schema to be consistent throughout the doc.
- Completely rewrote the Reference Object and updated its examples. I didn't realize the wording was relevant to the 2.0 but not 3.0. The wording should be simpler now, explain exactly what it is or isn't, and avoid the potential confusion regarding its resolution.

@tdc - please review again.

darrelmiller commented:
Yes, the wording is even better now. :shipit:

whitlockjc commented on versions/3.0.md:
```diff
@@ -2287,25 +2284,32 @@ The following properties are taken directly from the JSON Schema definition and
+- not - Inline or referenced schema MUST be of a [Schema Object](#schemaObject) and not a standard JSON Schema.
+- items - Value MUST be an object and not an array. Inline or referenced schema MUST be of a [Schema Object](#schemaObject) and not a standard JSON Schema. `items` MUST be present if the `type` is `array`.
+- properties - Property definitions MUST be a [Schema Object](#schemaObject) and not a standard JSON Schema (inline or referenced).
+- additionalProperties - Value can be boolean or object. If object, definition MUST be a [Schema Object](#schemaObject) and not a standard JSON Schema (inline or referenced).
```
Can it be a _JSON Reference_ to a _Schema Object_?

whitlockjc commented on versions/3.0.md:
```diff
@@ -2287,25 +2284,32 @@ The following properties are taken directly from the JSON Schema definition and
+- items - Value MUST be an object and not an array. Inline or referenced schema MUST be of a [Schema Object](#schemaObject) and not a standard JSON Schema. `items` MUST be present if the `type` is `array`.
+- properties - Property definitions MUST be a [Schema Object](#schemaObject) and not a standard JSON Schema (inline or referenced).
+- additionalProperties - Value can be boolean or object. If object, definition MUST be a [Schema Object](#schemaObject) and not a standard JSON Schema (inline or referenced).
+- $ref - A [Reference Object](#referenceObject).
```
Should we mention this at all?  To me, any place you could have a _Schema Object_ you would allow a _JSON Reference_ to a _Schema Object_ but it is an either/or, as in you cannot mix a _JSON Reference_ and a _Schema Object_ in the same object.  I think we could have some extra wording here making sure this is known to be the case, much like we document the `items` property as being applicable and required only when `type: array`.

whitlockjc commented on versions/3.0.md:
```diff
@@ -2287,25 +2284,32 @@ The following properties are taken directly from the JSON Schema definition and
+- $ref - A [Reference Object](#referenceObject).
+- description - [CommonMark syntax](http://spec.commonmark.org/) can be used for rich text representation.
+- format - See [Data Type Formats](#dataTypeFormat) for further details. While relying on JSON Schema's defined formats, the OAS offers a few additional predefined formats.
+- default - Unlike JSON Schema, the value MUST conform to the defined type for the Schema Object.
```
You don't specifically tell what the type is, or its purpose.  Would it make sense to do this and then throw the existing content into a _"(Note: ...)"_?

whitlockjc commented on versions/3.0.md:
```diff
@@ -2323,7 +2327,9 @@ While composition offers model extensibility, it does not imply a hierarchy betw
 When used, the `discriminator` will be the name of the property used to decide which schema definition is used to validate the structure of the model.
 As such, the `discriminator` field MUST be a required field.
-The value of the chosen property has to be the friendly name given to the model under the `definitions` property.
+There are are two ways to define the value of a discriminator for an inheriting instance.
```
Without an example, this portion is unclear to me.

whitlockjc replied:
Actually, I _might_ get this.  Are you saying that the value of the `discriminator` field must match the name of a definition?  Does this mean you can **only** use values that match keys of `#/components/definitions`?

DavidBiesack replied:
Good point -- we rely heavily on external schema files via `$ref`. For example, if Cat and Dog are defined externally, I should still be able to use them. However for a schema included from an external schema.json there is no name as there is for one defined in `/components/definitions`. If all three schema (Pet, Cat, Doc) were defined externally (so that they can be used in multiple OpenAPI documents) and there is no single `/components/definitions` in which they reside, how would this work?

webron replied:
We'll be handling the `discriminator` in a separate PR.

DavidBiesack commented on versions/3.0.md:
```diff
@@ -2323,7 +2327,9 @@ While composition offers model extensibility, it does not imply a hierarchy betw
-The value of the chosen property has to be the friendly name given to the model under the `definitions` property.
+There are are two ways to define the value of a discriminator for an inheriting instance.
+- Use the definition's name.
+- Override the definition's name by overriding the property with a new value. If exists, this takes precedence over the definition's name.
```
I'm not sure what this means. Override which property - the discriminator field (`petType` in this example)? If so, how is it overridden and specified? Would I do:

```YAML
  Cat:  ## "Cat" will be used as the discriminator value
    description: A representation of a cat
    allOf:
    - $ref: '#/definitions/Pet'
    - type: object
      properties:
         petType:
              type: string
              enum:
              - Cat
          huntingSkill:
             ...
```

An example would help.

webron commented:
Thanks @DavidBiesack - I'd still rather leave the wording regarding the JSON Schema intact, based on my experience with users in the past.

darrelmiller commented:
Latest updates LGTM 🎈

whitlockjc commented on versions/3.0.md:
```diff
@@ -2253,25 +2254,21 @@ $ref: 'Pet.yaml'
+This object is an extended subset of the [JSON Schema Specification Wright Draft 00](http://json-schema.org/).
 
-Further information about the properties can be found in [JSON Schema Core](http://json-schema.org/latest/json-schema-core.html) and [JSON Schema Validation](http://json-schema.org/latest/json-schema-validation.html).
+Further information about the properties can be found in [JSON Schema Core](https://tools.ietf.org/html/draft-wright-json-schema-00) and [JSON Schema Validation](https://tools.ietf.org/html/draft-wright-json-schema-validation-00).
```
Here you call it _"JSON Schema Core"_ but you reference it earlier as _"JSON Schema Specification Wright Draft 00"_.  Make a decision and use it in both places.

webron replied:
JSON Schema is split into two specifications - Core and Validation. Can't keep it as one, and shouldn't reference both every single time.

whitlockjc replied:
Fair enough.

whitlockjc commented on versions/3.0.md:
```diff
@@ -2287,25 +2284,33 @@ The following properties are taken directly from the JSON Schema definition and
+- format - See [Data Type Formats](#dataTypeFormat) for further details. While relying on JSON Schema's defined formats, the OAS offers a few additional predefined formats.
+- default - The default value represents what would be assumed by the consumer of the input as the value of the schema if one is not provided. Unlike JSON Schema, the value MUST conform to the defined type for the Schema Object defined at the same level. For example, of `type` is `string`, then `default` can be `"foo"`` but cannot be `1`.
+
+Alternatively, any time a Scheme Object can be used, a [Reference Object](#referenceObject) can be used in place. This allows referencing definitions in place of defining them inline. Effectively, it is the `$ref` property in JSON Schema but follows the same guidelines and restrictions as the Reference Object. For example, additional properties alongside a `$ref` SHALL be ignored.
```
Should _"in place"_ be _"in its place"_?

whitlockjc replied:
The third sentence doesn't make sense to me.  What about mentioning the caveat about mixing other properties with `$ref` into the _Reference Object_ documentation and omitting the last two sentences?

whitlockjc commented:
:shipit:

webron commented:
:tada:
